content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+  metrics from Kubelet in real-time. It then returns these metrics to you. Since the information
+  is generated in real-time and is only relevant at the moment of the request, there is no need
+  to store it in etcd. This approach saves resources.
+
+* If needed, you can use a backend other than etcd. You can even implement a Kubernetes-compatible API
+  for it. For example, if you use Postgres, you can create a transparent representation of its entities
+  in the Kubernetes API. Eg. databases, users, and grants within Postgres would appear as regular
+  Kubernetes resources, thanks to your extension api-server. You cloud manage them using `kubectl` or any
+  other Kubernetes-compatible tool. Unlike operators, which implement business logic using custom resources
+  and reconciliation methods, an api-server eliminates the need for separate controllers for every kind.
+  This means you don't have to sync state between Kubernetes and your backend.
+  Instead Kubernetes will expose it directly.
+
+### One-Time resources
+
+- Kubernetes has a special API used to provide users with information about their permissions.
+  This API is implemented using `SelfSubjectAccessReview` resources. The specificity of these
+  resources is that you can't view them using GET or LIST methods. You can only create them using
+  the CREATE method and receive output with information about what you have access to at the
+  moment.
+  
+  {{% alert title="Note" color="info" %}}
+  If you try to run `kubectl get selfsubjectaccessreviews` directly, you'll just get an error
+  like this:
+  
+  ```shell
+  Error from server (MethodNotAllowed): the server does not allow this method on the requested resource
+  ```
+  
+  The reason is that the Kubernetes api-server doesn't support any other interaction with this
+  type of resource (you can only CREATE them).
+  
+  This API, implemented using `SelfSubjectAccessReview` resources, is used for commands like:
+  
+  ```shell
+  # kubectl auth can-i create deployments --namespace dev
+  ```
+  
+  When you run the command above, Kubernetes creates a `SelfSubjectAccessReview` resource in the
+  Kubernetes API. This allows Kubernetes to fetch a list of possible permissions for your user.
+  Kubernetes then generates a personalized response to your request in real-time. This logic is
+  different from a scenario where this resource is simply stored in etcd.
+  {{% /alert %}}
+
+- Similarly, in KubeVirt's [CDI (Containerized Data Importer)](https://github.com/kubevirt/containerized-data-importer)
+  extension, which allows file uploads into a PVC from a local machine using the `virtctl` tool,
+  a special token is required before the upload process begins.
+  This token is generated by creating an `UploadTokenRequest` resource in the Kubernetes API. Kubernetes
+  routes all `UploadTokenRequest` resources creation requests to the CDI api-server,
```

Comment:
``suggestion
  This token is generated by creating an UploadTokenRequest resource via the Kubernetes API. Kubernetes
  routes (proxies) all UploadTokenRequest resource creation requests to the CDI extension API server,
```